openshift-knative-operator/pkg/monitoring/monitoring.go

```diff
@@ -0,0 +1,115 @@
+package monitoring
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	mf "github.com/manifestival/manifestival"
+	"github.com/openshift-knative/serverless-operator/openshift-knative-operator/pkg/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	EnableMonitoringEnvVar       = "ENABLE_SERVING_MONITORING_BY_DEFAULT"
+	EnableMonitoringLabel        = "openshift.io/cluster-monitoring"
+	ObservabilityCMName          = "observability"
+	ObservabilityBackendKey      = "metrics.backend-destination"
+	OpenshiftMonitoringNamespace = "openshift-monitoring"
+)
+
+const (
+	prometheusRoleName        = "knative-serving-prometheus-k8s"
+	prometheusClusterRoleName = "rbac-proxy-metrics-prom"
+)
+
+func SetupServingMonitoring(ctx context.Context, api kubernetes.Interface, ks *v1alpha1.KnativeServing) error {
+	backend, isSet := ks.Spec.CommonSpec.Config[ObservabilityCMName][ObservabilityBackendKey]
+	if shouldEnableMonitoring(backend) {
+		log := logging.FromContext(ctx)
+		log.Info("Enabling Serving monitoring")
+		if err := setMonitoringLabelToNamespace(ctx, ks.Namespace, api, true); err != nil {
+			return fmt.Errorf("failed to enable monitoring %w ", err)
+		}
+		return nil
+	}
+	if err := setMonitoringLabelToNamespace(ctx, ks.Namespace, api, false); err != nil {
+		return err
+	}
+	if !isSet {
+		common.Configure(&ks.Spec.CommonSpec, ObservabilityCMName, ObservabilityBackendKey, "none")
+	}
+	return nil
+}
+
+func shouldEnableMonitoring(backend string) bool {
+	enable, present := os.LookupEnv(EnableMonitoringEnvVar)
+	// Skip setup from env if feature toggle is off, use whatever the user defines in the Serving CR.
+	if !present && backend != "none" {
+		return true
+	}
+	parsedEnable, _ := strconv.ParseBool(enable)
```

Comment:
Swallowing the error feels a little odd. What I've seen other people do is to compare with "true", i.e.: `enable := strings.EqualFold("true", enable)`.

I'd actually also generalize the !present case above to that as well, i.e.

```golang
func shouldEnableMonitoring(backend string) bool {
  enableByDefault := strings.EqualFold(os.GetEnv(EnableMonitoringEnvVar), "true")
  if !enableByDefault && backend != "none" && backend != "" {
    return true
  }
  return enableByDefault
}
```

Is imo easier to understand.

Reply:
Yes but `  enableByDefault := strings.EqualFold(os.GetEnv(EnableMonitoringEnvVar), "true")` does not cover the case where no env var is set and backend is not `none`. If var is not set and backend is not `none` I need to return true.
`EqualFold` does not help with that.

Reply:
I changed it a bit.